# Pandas Practice 1
저번에 썼던 `student.csv`와 `Pandas`를 이용해서 간단한 데이터 전처리를 해봅시다!

In [1]:
import pandas as pd

### Q1. `student.csv` 파일을 불러와 `df` 변수에 저장해봅시다.

In [2]:
df = pd.read_csv('student.csv')
df.head()

id  present  absent  late  midterm  final  hw1  hw2 course_retake
0  20162493       32       0     0       60     20   10   10             Y
1  20182843       20      10     2       75     28   10   10           NaN
2  20173923       10       2    20       24     24   10    7           NaN
3  20191100       30       1     1       87     29   10    8           NaN
4  20132011       31       0     1       56     31    8    7           NaN

### Q2. DataFrame에서 앞의 7개의 row를 출력해봅시다.

In [3]:
df.head(7)

id  present  absent  late  midterm  final  hw1  hw2 course_retake
0  20162493       32       0     0       60     20   10   10             Y
1  20182843       20      10     2       75     28   10   10           NaN
2  20173923       10       2    20       24     24   10    7           NaN
3  20191100       30       1     1       87     29   10    8           NaN
4  20132011       31       0     1       56     31    8    7           NaN
5  20150029       29       3     0       93     19    9    6             Y
6  20163921        1      30     1       32     43   10   10           NaN

### Q3. 학생들의 수는 총 몇 명인가요?    
21명

In [4]:
df.shape[0]

21

### Q4. 값이 채워지지 않은, NULL값이 있는 column의 이름은 무엇인가요?
course_retake

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 9 columns):
id               21 non-null int64
present          21 non-null int64
absent           21 non-null int64
late             21 non-null int64
midterm          21 non-null int64
final            21 non-null int64
hw1              21 non-null int64
hw2              21 non-null int64
course_retake    5 non-null object
dtypes: int64(8), object(1)
memory usage: 1.6+ KB


### Q5. `id` column에서 학생의 학번을 뽑아서 `admin_year`라는 column을 생성해봅시다. 
* 힌트1 : 위의 문제에서 `id`의 자료형을 확인해봅시다.
* 힌트2 : 내가 아는 연산자를 한 번 사용해봅시다. (람다식 ㄴㄴ)

In [6]:
df['admin_year'] = df['id']//10000
df.head()

id  present  absent  late  midterm  final  hw1  hw2 course_retake  \
0  20162493       32       0     0       60     20   10   10             Y   
1  20182843       20      10     2       75     28   10   10           NaN   
2  20173923       10       2    20       24     24   10    7           NaN   
3  20191100       30       1     1       87     29   10    8           NaN   
4  20132011       31       0     1       56     31    8    7           NaN   

   admin_year  
0        2016  
1        2018  
2        2017  
3        2019  
4        2013

### Q6. 각 학번 당 몇 명이 있나요?

In [7]:
df['admin_year'].value_counts()

2018    6
2017    5
2016    4
2019    3
2015    1
2014    1
2013    1
Name: admin_year, dtype: int64

### Q7. 학생 id가 `20173923`인 학생이 중도휴학을 신청했습니다. 해당 학생은 점수 계산시 총점이 들어가지 않는다고 하네요. 

#### Q7-1. 해당 학생의 데이터를 DataFrame에서 지워봅시다.
* 참고 : id가 `20173923`인 학생의 index는 2입니다.
* `inplace` : True면, 변경 내역을 원래 DataFrame에 반영

In [8]:
df.drop([2], axis=0, inplace=True)
df.head()

id  present  absent  late  midterm  final  hw1  hw2 course_retake  \
0  20162493       32       0     0       60     20   10   10             Y   
1  20182843       20      10     2       75     28   10   10           NaN   
3  20191100       30       1     1       87     29   10    8           NaN   
4  20132011       31       0     1       56     31    8    7           NaN   
5  20150029       29       3     0       93     19    9    6             Y   

   admin_year  
0        2016  
1        2018  
3        2019  
4        2013  
5        2015

#### Q7-2. DataFrame의 index를 추출하여 출력해봅시다. index가 연속된 정수로 되어있나요?

In [9]:
df.index

Int64Index([0, 1, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
            20],
           dtype='int64')

#### Q7-3. index를 다시 연속된 정수의 형태로 바꿔줍시다. 기존의 index는 DataFrame에 포함하지 않습니다.
* `inplace` : True면, 기존 DataFrame에 반영
* `drop` : True면, 기존 index를 DataFrame에 포함시키지 않음

In [10]:
df.reset_index(drop=True, inplace=True)
df.head()

id  present  absent  late  midterm  final  hw1  hw2 course_retake  \
0  20162493       32       0     0       60     20   10   10             Y   
1  20182843       20      10     2       75     28   10   10           NaN   
2  20191100       30       1     1       87     29   10    8           NaN   
3  20132011       31       0     1       56     31    8    7           NaN   
4  20150029       29       3     0       93     19    9    6             Y   

   admin_year  
0        2016  
1        2018  
2        2019  
3        2013  
4        2015

### Q8. 출석 점수를 제외하고 반영 비율을 맞추어 총점을 계산하여 `total_ex_att`이라는 column을 생성하여 저장해봅시다.
* 중간고사 : 30% (100점 만점)
* 기말고사 : 40% (100점 만점)
* 과제1 : 10% (10점 만점)
* 과제2 : 10% (10점 만점)

In [11]:
df['total_ex_att'] = df['midterm']*0.3 + df['final']*0.4 + df['hw1'] + df['hw2']
df.head()

id  present  absent  late  midterm  final  hw1  hw2 course_retake  \
0  20162493       32       0     0       60     20   10   10             Y   
1  20182843       20      10     2       75     28   10   10           NaN   
2  20191100       30       1     1       87     29   10    8           NaN   
3  20132011       31       0     1       56     31    8    7           NaN   
4  20150029       29       3     0       93     19    9    6             Y   

   admin_year  total_ex_att  
0        2016          46.0  
1        2018          53.7  
2        2019          55.7  
3        2013          44.2  
4        2015          50.5

In [12]:
df.to_csv('student2.csv', mode='w')

### Q9. 계산하는데 사용한 지표들을 삭제해봅시다.
* midterm, final, hw1, hw2
* 원래 DataFrame에 변경 내용을 적용합니다.

In [13]:
df.drop(['midterm', 'final', 'hw1', 'hw2'], axis=1, inplace=True)
df.head()

id  present  absent  late course_retake  admin_year  total_ex_att
0  20162493       32       0     0             Y        2016          46.0
1  20182843       20      10     2           NaN        2018          53.7
2  20191100       30       1     1           NaN        2019          55.7
3  20132011       31       0     1           NaN        2013          44.2
4  20150029       29       3     0             Y        2015          50.5

In [14]:
df.to_csv('result_ex_att.csv', mode='w')